src/particles/CollectLost.cpp

```diff
@@ -0,0 +1,150 @@
+/* Copyright 2022-2023 The Regents of the University of California, through Lawrence
+ *           Berkeley National Laboratory (subject to receipt of any required
+ *           approvals from the U.S. Dept. of Energy). All rights reserved.
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Axel Huebl, Chad Mitchell
+ * License: BSD-3-Clause-LBNL
+ */
+#include "CollectLost.H"
+
+#include <AMReX_GpuLaunch.H>
+#include <AMReX_GpuQualifiers.H>
+#include <AMReX_Math.H>
+#include <AMReX_ParticleTransformation.H>
+#include <AMReX_RandomEngine.H>
+
+
+namespace impactx
+{
+    void collect_lost_particles (ImpactXParticleContainer& source, ImpactXParticleContainer& dest)
+    {
+        BL_PROFILE("impactX::collect_lost_particles");
+
+        using SrcData = ImpactXParticleContainer::ParticleTileType::ConstParticleTileDataType;
+        using DstData = ImpactXParticleContainer::ParticleTileType::ParticleTileDataType;
+
+        RefPart const ref_part = source.GetRefParticle();
+        auto const s_lost = ref_part.s;
+
+        // copy all particles marked with a negative ID from source to destination
+        int const nLevel = source.finestLevel();
+        for (int lev = 0; lev <= nLevel; ++lev) {
+            // loop over all particle boxes
+            using ParIt = ImpactXParticleContainer::iterator;
+            auto& plevel = source.GetParticles(lev);
+
+            // TODO: is it safe to add OpenMP parallelism here?
+            for (ParIt pti(source, lev); pti.isValid(); ++pti) {
+                auto index = std::make_pair(pti.index(), pti.LocalTileIndex());
+                if (plevel.find(index) == plevel.end()) continue;
+
+                auto& ptile_source = plevel.at(index);
+                auto const np = ptile_source.numParticles();
+                if (np == 0) continue;  // no particles in source tile
+
+                // we will copy particles that were marked as lost, with a negative id
+                auto predicate = [](const SrcData& src, int ip, const amrex::RandomEngine& /*engine*/) noexcept
+                {
+                    return src.id(ip) < 0;
+                };
+
+                // FIXME This seems to segfault
+                auto& ptile_dest = dest.DefineAndReturnParticleTile(
+                        lev, pti.index(), pti.LocalTileIndex());
```

Comment:
This seems to segfault.
Maybe the container is not properly initialized yet?

Reply:
Can you print out `source.finestLevel()` and `dest.finestLevel()`?

Reply:
Both fail for a non-MR sim on level 0:
```
src.finest=0
dest.finest=0
Segfault
```

Reply:
We found it - dependency issue between `AddRealComp`, `InitFromScratch` and `reserveData`, `resizeData`.